src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,358 @@
+                        {
+                            _workspaceHandlers.Remove(workspaceHandler);
+                        }
+
+                        foreach (var documentHandler in assemblyHandlers.DocumentMessageHandlers.Keys)
+                        {
+                            _documentHandlers.Remove(documentHandler);
+                        }
+                    }
+                }
+
+                if (extension.AssemblyHandlersCount > 0)
+                {
+                    return ValueTask.CompletedTask;
+                }
+
+                _extensions.Remove(assemblyFolderPath);
+            }
+        }
+
+        extension?.AnalyzerAssemblyLoader.Dispose();
+
+        return ValueTask.CompletedTask;
+    }
+
+    public void Reset()
+    {
+        List<Extension> extensions;
+        lock (_lockObject)
+        {
+            extensions = _extensions.Values.ToList();
+            _extensions.Clear();
+            _workspaceHandlers.Clear();
+            _documentHandlers.Clear();
+        }
+
+        foreach (var extension in extensions)
+        {
+            extension.AnalyzerAssemblyLoader.Dispose();
+        }
+    }
+
+    public void Dispose()
+        => Reset();
+
+    private async ValueTask<string> HandleExtensionMessageAsync<TArgument>(
+        TArgument argument,
+        string messageName,
+        string jsonMessage,
+        Dictionary<string, IExtensionMessageHandlerWrapper<TArgument>> handlers,
+        CancellationToken cancellationToken)
+    {
+        IExtensionMessageHandlerWrapper<TArgument> handler;
+        lock (_lockObject)
+        {
+            // handlers here is either _workspaceHandlers or _documentHandlers, so it must be protected
+            // by _lockObject.
+            if (!handlers.TryGetValue(messageName, out handler!))
+            {
+                throw new InvalidOperationException($"No handler found for message {messageName}.");
+            }
+        }
+
+        // Any exception thrown in this method is left to bubble up to the extension.
+        var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+        var result = await handler.ExecuteAsync(message, argument, cancellationToken)
+            .ConfigureAwait(false);
+        var responseJson = JsonSerializer.Serialize(result, handler.ResponseType);
+        return responseJson;
+    }
+
+    private void RegisterAssembly(Extension extension, string assemblyFileName, AssemblyHandlers? assemblyHandlers)
+    {
+        lock (_lockObject)
+        {
+            // Make sure a call to UnloadCustomMessageHandlersAsync hasn't happened while we relinquished the lock on _lockObject
+            if (!_extensions.TryGetValue(extension.AssemblyFolderPath, out var currentExtension) || !currentExtension.Equals(extension))
+            {
+                throw new InvalidOperationException($"Handlers in {extension.AssemblyFolderPath} were unregistered while loading handlers.");
```

Comment:
this feels more compelx than needed.  extension loading seems like it should be super rare.  so can we not just take the hit under the single lock and be ok?

Reply:
Tagging @tmat, since he gave the opposite feedback. I am ok with either solution (it's an implementation detail, so something that we can change later)

Reply:
i wouldn't mind starting simple as well, and reporting teleemetry on 'time to load extension' and 'time waiting for lock', so we can get data to drive the need to have complexity to allow interleaving.  we can always optimzie later, but be in a very clear and safe place to start.